File: services/golang/internal/service/import_service.go
package service

import (
	"context"
	"fmt"
	"log/slog"
	"sync"

	"github.com/google/uuid"
	"golang.org/x/sync/errgroup"
	"golang.org/x/time/rate"

	"github.com/igorrmotta/api-corestack/services/golang/internal/domain"
)

type ImportService struct {
	taskRepo    domain.TaskRepository
	notifRepo   domain.NotificationRepository
	concurrency int
	rateLimit   rate.Limit
}

func NewImportService(taskRepo domain.TaskRepository, notifRepo domain.NotificationRepository, concurrency int, rateLimit float64) *ImportService {
	if concurrency <= 0 {
		concurrency = 10
	}
	if rateLimit <= 0 {
		rateLimit = 100
	}
	return &ImportService{
		taskRepo:    taskRepo,
		notifRepo:   notifRepo,
		concurrency: concurrency,
		rateLimit:   rate.Limit(rateLimit),
	}
}

func (s *ImportService) BulkImport(ctx context.Context, workspaceID, projectID uuid.UUID, inputs []domain.TaskInput) *domain.ImportResult {
	result := &domain.ImportResult{
		Total: int32(len(inputs)),
	}

	if len(inputs) == 0 {
		return result
	}

	slog.InfoContext(ctx, "starting bulk import",
		"workspace_id", workspaceID,
		"project_id", projectID,
		"total", len(inputs),
		"concurrency", s.concurrency,
	)

	var mu sync.Mutex
	limiter := rate.NewLimiter(s.rateLimit, s.concurrency)

	g, ctx := errgroup.WithContext(ctx)
	g.SetLimit(s.concurrency)

	for i, input := range inputs {
		i, input := i, input // capture loop vars
		g.Go(func() error {
			if err := limiter.Wait(ctx); err != nil {
				mu.Lock()
				result.Failed++
				result.Errors = append(result.Errors, domain.ImportError{
					Index: int32(i),
					Error: fmt.Sprintf("rate limit: %v", err),
				})
				mu.Unlock()
				return nil // don't cancel other goroutines
			}

			task, err := s.taskRepo.Create(ctx, domain.CreateTaskParams{
				WorkspaceID: workspaceID,
				ProjectID:   projectID,
				Title:       input.Title,
				Description: input.Description,
				Priority:    input.Priority,
				AssignedTo:  input.AssignedTo,
				DueDate:     input.DueDate,
				Metadata:    input.Metadata,
			})
			if err != nil {
				mu.Lock()
				result.Failed++
				result.Errors = append(result.Errors, domain.ImportError{
					Index: int32(i),
					Error: err.Error(),
				})
				mu.Unlock()
				return nil
			}

			// Enqueue notification for each imported task
			if s.notifRepo != nil {
				_, _ = s.notifRepo.Create(ctx, domain.CreateNotificationParams{
					WorkspaceID: workspaceID,
					EventType:   "task.imported",
					Payload:     []byte(fmt.Sprintf(`{"task_id":"%s","title":"%s"}`, task.ID, task.Title)),
				})
			}

			mu.Lock()
			result.Succeeded++
			mu.Unlock()

			return nil
		})
	}

	_ = g.Wait()

	slog.InfoContext(ctx, "bulk import completed",
		"total", result.Total,
		"succeeded", result.Succeeded,
		"failed", result.Failed,
	)

	return result
}

File: services/typescript/src/service/import-service.ts
import PQueue from "p-queue";
import { NotificationRepo } from "../repository/notification-repo.js";
import {
  TaskRepo,
  type ImportError,
  type ImportResult,
  type TaskInput,
} from "../repository/task-repo.js";
import type { Logger } from "pino";

export class ImportService {
  private concurrency: number;
  private rateLimit: number;

  constructor(
    private taskRepo: TaskRepo,
    private notifRepo: NotificationRepo | null,
    private logger: Logger,
    concurrency = 10,
    rateLimit = 100,
  ) {
    this.concurrency = concurrency > 0 ? concurrency : 10;
    this.rateLimit = rateLimit > 0 ? rateLimit : 100;
  }

  async bulkImport(
    workspaceId: string,
    projectId: string,
    inputs: TaskInput[],
  ): Promise<ImportResult> {
    const result: ImportResult = {
      total: inputs.length,
      succeeded: 0,
      failed: 0,
      errors: [],
    };

    if (inputs.length === 0) {
      return result;
    }

    this.logger.info(
      { workspaceId, projectId, total: inputs.length, concurrency: this.concurrency },
      "starting bulk import",
    );

    const queue = new PQueue({
      concurrency: this.concurrency,
      interval: 1000,
      intervalCap: this.rateLimit,
    });

    const errors: ImportError[] = [];

    const promises = inputs.map((input, i) =>
      queue.add(async () => {
        try {
          const task = await this.taskRepo.create({
            workspaceId,
            projectId,
            title: input.title,
            description: input.description,
            priority: input.priority,
            assignedTo: input.assignedTo,
            dueDate: input.dueDate,
            metadata: input.metadata,
          });

          // Enqueue notification for each imported task
          if (this.notifRepo) {
            try {
              await this.notifRepo.create({
                workspaceId,
                eventType: "task.imported",
                payload: { task_id: task.id, title: task.title },
              });
            } catch {
              // fire-and-forget
            }
          }

          result.succeeded++;
        } catch (err) {
          result.failed++;
          errors.push({
            index: i,
            error: err instanceof Error ? err.message : String(err),
          });
        }
      }),
    );

    await Promise.all(promises);
    result.errors = errors;

    this.logger.info(
      { total: result.total, succeeded: result.succeeded, failed: result.failed },
      "bulk import completed",
    );

    return result;
  }
}
